scripts/processing/segment_sentences.py

```diff
@@ -0,0 +1,329 @@
+
+            def work(articles, return_dict, rank):
+                print('process {} needs to segment {} articles'.format(rank,len(articles)))
+                sentences = {}
+                for i, article in enumerate(articles):
+                    sentences[article] = segmenter.segment_string(articles[article])
+
+                    if i % 100 == 0:
+                        print('process {} finish article {}'.format(rank,i))
+
+                print('process {} finish segment'.format(rank))
+                return_dict.update(sentences)
+            rank = 0
+            for item in chunks(self.articles, len(self.articles)//(n_processes)+1):
+                p = multiprocessing.Process(target=work, args=(item, return_dict,rank))
+                rank+=1
+                # Busy wait
+                while len(jobs) >= n_processes:
+                    pass
+
+                jobs.append(p)
+                p.start()
+
+            for proc in jobs:
+                proc.join()
+            self.sentences.update(return_dict)
+
+            time2 = time.time()
+
+            print('total segment spend  spend:', time2-time1)
+        elif use_multiprocessing == 'queue':
+            work_queue = multiprocessing.Queue()
+            jobs = []
+
+            for item in chunks(self.articles, len(self.articles)):
+                pass
+
+        else:
+            # serial option
+            time1= time.time()
+            for i, article in enumerate(self.articles):
+                self.sentences[i] = segmenter.segment_string(self.articles[article])
+
+                if i % 100 == 0:
+                    print('Segmenting article', i)
+            time2 = time.time()
+            print('segmentation spend:', time2-time1)
+        print('End: Sentence Segmentation')
+
+
+    def init_output_files(self):
+        print('Start: Init Output Files')
+        assert len(self.output_training_files) is 0, 'Internal storage self.output_files already contains data. This function is intended to be used by the constructor only.'
+        assert len(self.output_test_files) is 0, 'Internal storage self.output_files already contains data. This function is intended to be used by the constructor only.'
+
+        for i in range(self.n_training_shards):
+            name = self.output_name_prefix + self.output_training_identifier + '_' + str(i) + self.output_file_extension
+            self.output_training_files[name] = []
+
+        for i in range(self.n_test_shards):
+            name = self.output_name_prefix + self.output_test_identifier + '_' + str(i) + self.output_file_extension
+            self.output_test_files[name] = []
+
+        print('End: Init Output Files')
+
+
+    def get_sentences_per_shard(self, shard):
+        result = 0
+        for article_id in shard:
+            result += len(self.sentences[article_id])
+
+        return result
+
+
+    def distribute_articles_over_shards(self):

```

Comment:
Is it necessary for this function to execute on the main process? Why can't each of the `segmenting_num_worker` not directly output data to the shards?